packages/pgsearch/client.js

```diff
@@ -296,54 +292,62 @@ class Batch {
   // This method does not need to recursively invalidate, because each
   // document stores a complete, rolled-up picture of which other
   // documents it references.
-  async _invalidations(schema, branch, read) {
-    await this.client._iterateThroughRows(
-      'select id, type from documents where expires < now()', [], async({ id, type }) => {
-        this._touched[`${type}/${id}`] = this._touchCounter++;
-      });
-    await this.client.query('delete from documents where expires < now()');
-    await this.client.docsThatReference(branch, Object.keys(this._touched), async (doc, refs) => {
-      let { type, id } = doc.data;
-
-      if (this._isInvalidated(type, id, refs)) {
-        let sourceId = schema.types.get(type).dataSource.id;
-        // this is correct because IF this document's data source is currently
-        // doing a replace-all operation, it was either already touched (so
-        // this code isn't running) or it's old (so it's correct to have a
-        // non-current nonce).
-        let nonce = 0;
-        let context = new DocumentContext({
-          schema,
-          branch,
-          type,
-          id,
-          sourceId,
-          generation: nonce,
-          upstreamDoc: doc,
-          read
+  async _invalidations() {
+    let branches = Object.keys(this._touched).map(key => {
+      log.info('_touched: %s', key);
+      return key.split('/')[0];
+    }).filter((v, i, a) => a.indexOf(v) === i);
+
+    for (let branch of branches) {
+      await this.client._iterateThroughRows(
+        'select id, type from documents where expires < now()', [], async({ id, type }) => {
+          this._touched[`${branch}/${type}/${id}`] = this._touchCounter++;
         });
+      await this.client.query('delete from documents where expires < now()');
+      await this.client.docsThatReference(branch, Object.keys(this._touched), async (doc, refs) => {
```

Comment:
I think we drop the `branch` argument to `docsThatReference` because it's now part of `this._touched` instead.